[[ Java FFI Support ]] Add Java FFI support to libfoundation and libscript

livecodeali opened the pull request:
Relies on https://github.com/livecode/livecode/pull/5161, so WIP until that is cleaned up and merged.
Supersedes https://github.com/livecode/livecode/pull/4529

livecodeali committed:
[[ Foundation ]] Add MCStringSubstringIsEqualToCString

livecodeali committed:
[[ Java FFI Support ]] Add support for java to travis Linux config
We need a Java runtime on Linux, and we need to be able to
dlopen the JLI shared library, so add its location to LD_LIBRARY_PATH

livecodeali committed:
[[ Java FFI Support ]] Add support for java to travis Mac config
Building the java portion of libfoundation requires the JAVA_HOME
environment variable to be set.

livecodeali committed:
[[ Java FFI Support ]] Find value for javahome gyp variable during config

livecodeali committed:
[[ SysLibrary ]] Allow direct use of system library API in weak linking
This patch adds a command-line option to the weak stub maker which allows
the generated code to load libraries directly using the foundation API
rather than going through the engine. This enables weak linking to libraries
within libfoundation.

livecodeali committed:
[[ Java FFI Support ]] Add foundation-java API
This patch adds support for calling Java methods via the JNI. An
MCJavaObjectRef opaque type has been added which wraps a jobject,
along with an API added to support manipulation of Java objects (eg
conversion to and from MCStringRefs) and constructing instances of
java classes.

livecodeali committed:
[[ Java FFI Support ]] Add Java FFI support to libscript
This patch adds support for Java foreign bindings in LCB via the
foreign binding string. Java bindings are specified using
"java:className>methodName(paramSignature)returnSignature", and
constructors using the methodName 'new'.

livecodeali committed:
[[ Java FFI Support ]] Add LCB Java support module
This patch adds an LCB java support module to libscript which wraps
the libfoundation Java support API. It attempts to initialise the
JVM when any of its functionality is used.

livecodeali committed:
[[ Java FFI ]] Add release notes and docs for java bindings

livecodeali committed:
[[ LCB Java ]] Add release note

peter-b commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -2964,16 +2972,28 @@ bool MCStringSubstringIsEqualToSubstring(MCStringRef self, MCRange p_sub, MCStri
-	__MCAssertIsString(self);
-	MCAssert(nil != p_chars);
+    return MCStringSubstringIsEqualToNativeChars(self,
+                                                 MCRangeMake(0, MCStringGetLength(self)),
```
For a very small efficiency improvement, you could use:

```c++
MCRangeMake(0,UINDEX_MAX)
```

here, because `MCStringSubstringIsEqualToNativeChars()` (and, transitively, `MCStringSubstringIsEqualTo()`) should already be clamping the range.

In fact it might make sense to add a `kMCEverythingRange` constant or something, since this a pretty frequent pattern.

peter-b commented on util/weak_stub_maker.pl:
```diff
@@ -83,9 +86,62 @@ sub trim
+	output "  t_symbol = MCSLibraryLookupSymbol(p_handle, t_symbol_name);";
+	output "  MCValueRelease(t_symbol_name);";
+	output "  return t_symbol;";
+	output "}";
```
This might be a bit easier to read (and edit) if you used a [Perl 'here document'](https://perlmaven.com/here-documents).

livecodeali replied:
I've added a bug report for this.
http://quality.livecode.com/show_bug.cgi?id=19334

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+#include <foundation.h>
```
Missing copyright header.

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+    // Just allow long and int interchangeably for now
+    if (t_code == kMCJavaTypeInt &&
+        p_code == kMCJavaTypeLong)
+        return true;
```
This is okay in one direction, but not in the other. Please file a bug report if you haven't already!

livecodeali replied:
http://quality.livecode.com/show_bug.cgi?id=19335

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+    return true;
+}
+
+static bool MCJavaTypeConforms(MCTypeInfoRef p_type, MCJavaType p_code)
```
It's not clear from the naming of this function or its parameters whether it's checking if `p_code` conforms to `p_type`, or _vice versa_ (or both!)

As far as I can tell, this is effectively `MCJavaTypeIsEquivalent()` or something.  Is that correct?

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+        case MCJavaCallTypeSetter:
+            return true;
+    }
+}
```
Please add an `MCUnreachableReturn(false)` at the end of this function, after the end of the `switch` statement.

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+    init_jvm_args(&vm_args);
+    
+    JavaVMOption* options = new (nothrow) JavaVMOption[1];
+    options[0].optionString = "-Djava.class.path=/usr/lib/java";
```
Is this `/usr/lib/java` path always the correct class path to pass here?

livecodeali replied:
Not in general, but we can add the ability to specify the user class path at a later date.
http://quality.livecode.com/show_bug.cgi?id=19336

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+    if (t_bytes == nullptr)
+        return false;
+
+    t_length = s_env -> GetArrayLength(p_byte_array);
```
`t_length` should be a `jsize`, and I think you need to check for overflow when converting to a `uindex_t` for the call to `MCDataCreateWithBytes` because `jsize` is 64 bits long on 64-bit platforms.  Might be a good idea to do a quick audit of JNI methods that return `jsize` to check for possible overflows.

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+    
+    const byte_t* t_data = MCDataGetBytePtr(p_data);
+    s_env -> SetByteArrayRegion(t_bytes, 0, t_length,
+                                reinterpret_cast<const jbyte*>(t_data));
```
To avoid a strict aliasing violation, I suggest writing this:

```c++
auto t_data = reinterpret_cast<const jbyte*>(MCDataGetBytePtr(p_data));
s_env->SetByteArrayRegion(t_bytes, 0, t_length, t_data);
```

I think that will avoid any possibility of the compiler deciding that you are accessing `t_data` via two distinct pointer types.

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+            MCUnreachableReturn(false);
+    }
+    return true;
+}
```
As we've discussed, these big `switch` could be simplified (and compressed!) with some suitable C++11 magic.

In the short term, it would be nice to get compilation warnings if the `kMCJavaType` enumeration changes.  Because there's a `default` label in the `switch`, we won't get that.

I suggest using this form:

```c++
switch (t_return_type)
{
    case kMCJavaTypeBoolean:
    {
        jboolean t_result =
            s_env -> CallBooleanMethodA(p_instance, p_method_id, p_params);
        *(static_cast<jboolean *>(r_result)) = t_result;
        return true;
    }

    /* ... other cases here */

    /* no default label */
}
MCUnreachableReturn(false);
```

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+    if (!MCJavaPrivateGetJObjectClassName(t_obj, &t_class_name))
+        return false;
+    
+    return MCStringFormat(r_desc, "<java: %@>", *t_class_name);
```
To be able to distinguish between different instances of the same Java class, I recommend including something unique (like the instance address) in the describe string.

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+            {
+                MCAutoStringRef t_sig;
+                if (!MCStringFormat(&t_sig, "%@%@", p_arguments, p_return))
+                    return nullptr;
```
`MCStringCreateWithStrings()` is a more efficient way to do this than `MCStringFormat()`.

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+                if (!MCStringCopySubstring(p_arguments,
+                                           MCRangeMake(1, MCStringGetLength(p_arguments) - 2),
+                                           &t_args))
+                    return nullptr;
```
This "get the arg string without its parentheses" operation is repeated in a couple of places, so consider splitting it out into a helper function.

peter-b commented on libfoundation/src/foundation-java.cpp:
```diff
@@ -0,0 +1,199 @@
+bool MCJavaCallJNIMethod(MCNameRef p_class_name, void *p_method_id, int p_call_type, MCTypeInfoRef p_signature, void *r_return, void **p_args, uindex_t p_arg_count)
+{
+    if (!s_java_initialised)
+        return false;
```
It might be a good idea to `MCErrorThrow()` something if code tries to do Java stuff without initialising Java. At the moment it will just fail silently.

peter-b commented on libfoundation/src/foundation-java-private.cpp:
```diff
@@ -0,0 +1,1353 @@
+            return type_map[i] . type;
+        }
+    }
+    
```
FYI, you can do some clever stuff with the STL to avoid the `sizeof()` ratio dance, e.g.:

```c++
#include <algorithm>

...

auto t_mapping = std::find(std::begin(type_map), std::end(type_map),
    [&](const java_type_map& p_entry) {
        return MCStringSubstringIsEqualToCString(p_type_code, p_range,
            p_entry.name, kMCStringOptionCompareExact);
    });
if (t_mapping != std::end(type_map))
    return t_mapping->type;
else
    return kMCJavaTypeObject;
```

peter-b commented on libscript/src/script-private.h:
```diff
@@ -314,14 +314,40 @@ struct MCScriptSyntaxDefinition: public MCScriptDefinition
+            void *method_id;
+            int call_type;
+        } java;
+    };
```
@livecodeali As you mentioned on Slack, it would be great to separate Java handler definitions from general foreign invocations by using a distinct storage type rather than turning `MCScriptForeignHandlerDefinition` into a union.

livecodeali replied:
http://quality.livecode.com/show_bug.cgi?id=19337

peter-b commented on libscript/src/module-java.cpp:
```diff
@@ -0,0 +1,105 @@
+        MCErrorCreateAndThrow(kMCGenericErrorTypeInfo, "reason",
+                              MCSTR("couldn't convert java object to string"),
+                              nullptr);
+}
```
I feel like it would be good to have distinct type infos for the the Java-specific errors rather than using the generic error type.

peter-b commented on docs/guides/LiveCode Builder Language Reference.md:
```diff
@@ -515,19 +532,40 @@ All but 'default' are Win32-only, and on Win32 'default' maps to
+
+ - instance
+ - static
+ - nonvirtual
```
I suggest putting these in `code markup`

peter-b commented on docs/guides/LiveCode Builder Language Reference.md:
```diff
@@ -515,19 +532,40 @@ All but 'default' are Win32-only, and on Win32 'default' maps to
+ - nonvirtual
+
+Instance and nonvirtual calling conventions require instances of the given
+java class, so the foreign handler declaration will always require a Java
```
"java" -> "Java"

peter-b commented on docs/lcb/notes/feature-java.md:
```diff
@@ -0,0 +1,15 @@
+## Java utilities
+
+There is now a utility library for manipulating java objects. It contains
+a type JObject which wraps a java object, some type conversion operations:
```
"java" -> "Java"
I suggest putting `JObject` (and other code symbols) in `code sample markup`

peter-b reviewed (commented):
Looking good! I've left some general feedback but I'm not sure what subset of it can be reasonably addressed in this iteration.

livecodeali committed:
[[ Java FFI ]] Remove superfluous break
We should never call GetFieldResult with a void return type.

livecodeali committed:
[[ Java FFI ]] Fix JNI env exception handling
Ensure JNI Env exceptions propagate as LCB errors.

livecodeali committed:
[[ Foundation ]] Use more efficient call to ranged string function

livecodeali committed:
[[ Java FFI ]] Add missing copyright header

livecodeali committed:
[[ Java FFI ]] Add missing MCUnreachableReturn

livecodeali committed:
[[ Java FFI ]] Clarify role of type info conformity helper

livecodeali committed:
[[ Java FFI ]] Fix issues with Data / JByteArray conversion

livecodeali committed:
[[ Java FFI ]] Tweak structure to enable compile enum warnings
Omitting the default case from switch statements ensures that the
compiler is able to detect missing cases if the enum is extended.

livecodeali committed:
[[ Java FFI ]] Output address of java object when describing
This helps to distinguish different instances of the same class.

livecodeali committed:
[[ Java FFI ]] Clean up some string manipulations

livecodeali committed:
[[ Java FFI ]] Fix styling in release notes

livecodeali committed:
[[ Java FFI ]] Fix styling in documentation

livecodeali committed:
[[ LCB Java ]] Add distinct typeinfos for Java-specific errors

livecodeali committed:
[[ Java FFI Support ]] Add suitable error typeinfos to foundation java

peter-b commented:
@livecode-vulcan review ok 9eaaef3

livecode-vulcan commented:
💙 review by @peter-b ok 9eaaef37384215ea652a037076cf32f50ab4071c

livecode-vulcan commented:
😎 test success 9eaaef37384215ea652a037076cf32f50ab4071c

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1005)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1011)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1003)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1004)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1011)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1025)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1002)
